### ISSUE TITLE
IconLayer crashing on Safari when updating layers and auto packing atlas

### ISSUE THREAD

**Issue by ruizmarc** [uber/deck.gl#3830]
#### Description

We are using IconLayer to represent live data with auto packing atlas (we generate a few hundred different icons depending on the data we receive). Everytime we receive an event with updated data, we set props with an updated layer and we get the changes represented in the map. 

Updating process is working fine on Chrome and Firefox, however, with Safari we only get crashes. 

It seems that it is a problem with WebGL2 not supported on Safari, but as first creation is working fine, it would be great to we able to update the layers the same we are creating them (creation works without problems) in the cases in which WebGL2 is not supported. 

#### Repro Steps

We update the layer, using set props like this:

`
this.overlay.setProps({ layers: [
      clustersLayer.clone({ data: clustersData })
]}); 
`

The amount of data entries may change depending on the bounds of the map in the screen and the events we receive, so we don't always have the same length of data (pointing this out, just in case it is relevant).

It is always working fine in Google Chrome and Firefox, but with Safari, altough it starts working fine (first creation is ok), once we update the layers with new data which generates new icons, it crashes. I'm attaching the stack trace. 

#### Environment (please complete the following information):
 - Framework Version: 7.3.3
 - Browser Version: Safari 13.0.3
 - OS: Mac OS Catalina 10.15.1

#### Logs

```
[Error] ERROR
Error: Requires WebGL2
assert — assert.js:3
assertWebGL2Context — webgl-checks.js:15
readPixelsToBuffer — copy-and-blit.js:78
resizeTexture — icon-manager.js:56
_updateAutoPacking — icon-manager.js:302
setProps — icon-manager.js:245
updateState — icon-layer.js:193
_updateState — layer.js:788
_update — layer.js:772
_updateLayer — layer-manager.js:501
_updateSublayersRecursively — layer-manager.js:383
_updateLayers — layer-manager.js:329
setLayers — layer-manager.js:189
setProps — layer-manager.js:148
setProps — deck.js:223
setProps — google-maps-overlay.js:48
updateOverlay — ng-components.js:4531
ngOnChanges — ng-components.js:4436
checkAndUpdateDirectiveInline — core.js:31905
debugCheckAndUpdateNode — core.js:45327
debugCheckDirectivesFn — core.js:45270
(anonymous function)
checkAndUpdateView — core.js:44270
callViewAction — core.js:44636
execEmbeddedViewsAction — core.js:44593
checkAndUpdateView — core.js:44271
callViewAction — core.js:44636
execComponentViewsAction — core.js:44564
checkAndUpdateView — core.js:44277
callViewAction — core.js:44636
execComponentViewsAction — core.js:44564
checkAndUpdateView — core.js:44277
callViewAction — core.js:44636
execEmbeddedViewsAction — core.js:44593
checkAndUpdateView — core.js:44271
callViewAction — core.js:44636
execComponentViewsAction — core.js:44564
checkAndUpdateView — core.js:44277
callWithDebugContext — core.js:45631
detectChanges — core.js:31125
tick — core.js:41044
(anonymous function) — core.js:40892
onInvoke — core.js:39698
run — zone-evergreen.js:124
next — core.js:40889
(anonymous function) — core.js:35335
__tryOrUnsub — Subscriber.js:183
next — Subscriber.js:122
_next — Subscriber.js:72
next — Subscriber.js:49
next — Subject.js:39
emit — core.js:35297
checkStable — core.js:39641
onHasTask — core.js:39718
hasTask — zone-evergreen.js:411
_updateTaskCount — zone-evergreen.js:431
_updateTaskCount — zone-evergreen.js:264
runTask — zone-evergreen.js:185
drainMicroTaskQueue — zone-evergreen.js:559
promiseReactionJob
	error
	View_MapComponent_1 (MapComponent.ngfactory.js:7)
	logError (vendor.js:99488)
	logError
	handleError (vendor.js:64138)
	run (polyfills.js:3359)
	tick (vendor.js:95185)
	(anonymous function) (vendor.js:95023)
	onInvoke (vendor.js:93957)
	run (polyfills.js:3359)
	next (vendor.js:95020)
	(anonymous function) (vendor.js:90285)
	__tryOrUnsub (vendor.js:257900)
	next (vendor.js:257839)
	_next (vendor.js:257789)
	next (vendor.js:257766)
	next (vendor.js:257552)
	emit (vendor.js:90247)
	checkStable (vendor.js:93900)
	onHasTask (vendor.js:93977)
	hasTask (polyfills.js:3646)
	_updateTaskCount (polyfills.js:3666)
	_updateTaskCount (polyfills.js:3499)
	runTask (polyfills.js:3420)
	drainMicroTaskQueue (polyfills.js:3794)
	promiseReactionJob
```

**Comment by 1chandu** [comment]
@ruizmarc  thanks for reporting, yes `readPixelsToArray` uses WebGL2 only API, we should replace this with `copyToTexture` , should be easy to update and happy to review if you can create a PR 
copyToTexture:  [doc](https://github.com/uber/luma.gl/blob/master/docs/api-reference/webgl/copy-and-blit.md#copytotexturesource--framebuffertexture-target-texturegl-enum-options-object--texture) and [source](https://github.com/uber/luma.gl/blob/51e1b7c8b298bc02114d73dab0757da38676e450/modules/webgl/src/classes/copy-and-blit.js#L168)

**Comment by ruizmarc** [comment]
Hi!

Thanks for the suggestion. I will work on the pull request. I'm not sure if I'm walking to the right direction because it seems it works better than before, but I'm having a new error. I'm attaching it at the end the comment. 

This is what I've understood from your suggestion about modifying resizeTexture function (let me know if this is what you were thinking about): 

```
function resizeTexture(texture, width, height) {
  const oldWidth = texture.width;
  const oldHeight = texture.height;
  const tempTexture = new Texture2D(texture.gl, { width: oldWidth, height: oldHeight});
  copyToTexture(texture, tempTexture);
  texture.resize({width, height});
  copyToTexture(tempTexture, texture, {
    targetY: height - oldHeight,
    height: height,
    width: width
  });
  texture.generateMipmap();
  return texture;
}
```

The error I'm having now is this one. Sorry for asking, but it is the first time working with WebGL and I need a few clues to make this pull request happen:

![image](https://user-images.githubusercontent.com/5717082/67957750-5e0dcc80-fbf6-11e9-94f3-a65c6ca5f948.png)

Thanks for your feedback

**Comment by 1chandu** [comment]
Is new width and height are always greater than current texture width and height?

I would try this :
```
function resizeTexture(texture, width, height) {
  const newTexture = cloneTextureFrom(texture, {width, height});
  copyToTexture(texture, newTexture);
  texture.delete();
  return newTexture;
}
```
also you need to update this._texture explictly.

**Comment by ruizmarc** [comment]
I have just tried this, and although is not showing any error in the console, at some point after a few data updates, icons stop being represented or just a few of them get represented. It is like it matches with empty pixels in the atlas. 

I have also tried your approach, but with the offset that was previously being applied, but I get similar results:

```
function resizeTexture(texture, width, height) {
  var oldHeight = texture.height;
  const newTexture = cloneTextureFrom(texture, {width, height});
  copyToTexture(newTexture, newTexture, {
    targetY: height - oldHeight,
    height: height,
    width: width
  });
  texture.delete();
  return newTexture;
}
```
With this offset, they also eventually stop being represented or just a few of them get represented. Furthermore, with the offset I also get the error I was mentioning in my previous comment. 

This is the evolution of resizes, just in case it may help:

```
[Log] oldHeight – 512 – "oldWidth" – 1024 – "newHeight" – 1024 – "newWidth" – 1024 (vendor.js, line 142583)
[Log] oldHeight – 1024 – "oldWidth" – 1024 – "newHeight" – 2048 – "newWidth" – 1024 (vendor.js, line 142583)
[Log] oldHeight – 2048 – "oldWidth" – 1024 – "newHeight" – 4096 – "newWidth" – 1024 (vendor.js, line 142583)

```

**Comment by Pessimistress** [comment]
Fix is landed in 7.3.5

**Comment by ruizmarc** [comment]
Yeah! Works great on Safari! Thanks a lot!
